teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,484 @@
+
+## Alternatives
+
+### Alternatives for Schemas
+
+Users of Tekton Chains need to have two results, `ARTIFACT_INPUTS` and `ARTIFACT_OUTPUTS`, that are arrays of strings which stores the signable artifacts' result names:
+
+```yaml
+results:
+  - name: ARTIFACT_INPUTS
+    value: ["ARTIFACT-INPUT-NAME-1", "ARTIFACT-INPUT-NAME-1"]
+  - name: ARTIFACT_OUTPUTS
+    value: ["ARTIFACT-OUTPUT-NAME-1"]
+```
+
+And the signable target results need to follow the below schema:
+
+```yaml
+results:
+  - name: ARTIFACT-NAME
+    type: object
+    description: |
+      * uri: resource uri of the artifact. It can uniquely identify the artifact.
+      * digest: revision digest in form algorithm:digest.
+    properties:
+      uri:
+        type: string
+      digest:
+        type: string
+```
+
+This design has been rejected because it would leave too much responsibilities for Task authors to write extra two results, which is not a good idea in real life practice.
+
+### Inputs / Outputs Distinguishment
+
+We can also separately collect inputs provenance from params, and outputs provenance from TaskRun results. The signable artifacts being found in params will be the inputs, and those found in TaskRun results will be the outputs. However, params are not reliable as users don't really specify those human less readable provenance metadata formats mentioned in in-toto attestations in neither PipelineRun nor TaskRun params.
+
+### Using Run Status to generate Provenance Metadata
+
+Results is not an ideal place to have the provenance metadata populated for the artifacts, since Results are provided by Pipeline authors. When people are looking for provenance, the metadata being provided should ideally be trustful and unchangable after generation. How can we garantee the provenance metadata can be trusted? Assuming the Tekton Pipeline being installed and where the Runs are operated are trustful, one missing piece to comply to SLSA level L3 is that the Run yaml being submitted is a trustful config, which can be fulfilled by [TEP 091: Trusted Resource](https://github.com/tektoncd/community/pull/739). The other missing piece is that we want to make sure the generated provenance data shouldn't be changable by any untrusted threads. TaskRun and PipelineRun `status` field satisfies these needs, as only Pipeline controller is able to generate info within this field. To extend the previous trustful setup, we can allow Trusted Tasks to modify the field as well: when Trusted Tasks are being used, Pipeline Controller can verify the Trusted Tasks being used in a Run, and let Trusted Tasks to generate provenance metadata a new field under `status` fields.
```

Comment:
Ah I brought up Trusted resources because I wanted to mention that currently in Tekton Pipelines, only the controller can change the `status` field. I reworded it a bit and shifted my main argument to be that we don't need to predefine a schema in Tasks. PTAL thanks!